src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs

```diff
@@ -117,50 +118,171 @@ private static string GetCurrentDirectoryForConditionEvaluation(ProjectElement e
 
         public struct ItemData
         {
-            private I _item;
-            private int _elementOrder;
-            private bool _conditionResult;
+            public ItemData(I item, ProjectItemElement originatingItemElement, int elementOrder, bool conditionResult)
+            {
+                Item = item;
+                OriginatingItemElement = originatingItemElement;
+                ElementOrder = elementOrder;
+                ConditionResult = conditionResult;
+            }
 
-            public ItemData(I item, int elementOrder, bool conditionResult)
+            public ItemData Clone(IItemFactory<I, I> itemFactory, ProjectItemElement initialItemElementForFactory)
             {
-                _item = item;
-                _elementOrder = elementOrder;
-                _conditionResult = conditionResult;
+                // setting the factory's item element to the original item element that produced the item
+                // otherwise you get weird things like items that appear to have been produced by update elements
+                itemFactory.ItemElement = OriginatingItemElement;
+                var clonedItem = itemFactory.CreateItem(Item, OriginatingItemElement.ContainingProject.FullPath);
+                itemFactory.ItemElement = initialItemElementForFactory;
+
+                return new ItemData(clonedItem, OriginatingItemElement, ElementOrder, ConditionResult);
             }
 
-            public I Item { get { return _item; } }
-            public int ElementOrder { get { return _elementOrder; } }
-            public bool ConditionResult { get { return _conditionResult; } }
+            public I Item { get; }
+            public ProjectItemElement OriginatingItemElement { get; }
+            public int ElementOrder { get; }
+            public bool ConditionResult { get; }
         }
 
+        private class MemoizedOperation : IItemOperation
+        {
+            private readonly IItemOperation _operation;
+            private Dictionary<ISet<string>, ImmutableList<ItemData>> _cache;
+
+            private bool _isReferenced;
+#if DEBUG
+            private bool _applyCalled;
+#endif
+
+            public MemoizedOperation(IItemOperation operation)
+            {
+                _operation = operation;
+            }
+
+            public void Apply(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore)
+            {
+#if DEBUG
+                Debug.Assert(!_applyCalled, "Apply should only be called once. Otherwise caching is not working");
```

Comment:
It should be valid for this method to be called multiple times if `globsToIgnore` is different each time.

(I left this comment on an outdated version of the file, so repeating it here in case that doesn't show up).